### DISCUSSION TITLE
Move cancel/save buttons to top of form

### DISCUSSION THREAD

**Discussion by lucacastelnuovo** [filamentphp/filament#6632] [Help] [answered]
Is it possible to move the save/cancel buttons at `marker 1` to `marker 2`?

I do not care if the `Delete` button then has to move down or it can be next to it

<img width="1202" alt="image" src="https://github.com/filamentphp/filament/assets/26206253/2bc16a7c-907b-43eb-83f3-ab1096be98da">

**Comment by RX-77IT** [comment]
I just tried this in the `EditResource.php` and it seems to work the way you want:

```php
use Filament\Pages\Actions;

// ...

    protected function getActions(): array
    {
        return [
            Actions\Action::make('save_top')
                ->action('save'),
            Actions\Action::make('cancel_top')
                ->url($this->getResource()::getUrl('index'))
                ->color('secondary'),
        ];
    }

    protected function getFormActions(): array
    {
        return []; // necessary to remove the bottom actions
    }
```

Documentation that I used: https://filamentphp.com/docs/2.x/admin/resources/editing-records#custom-actions

**Reply by lucacastelnuovo** [reply]
You are awesome. Will test this now!

**Reply by lucacastelnuovo** [reply, marked as answer]
Inspired by @RX-77IT solution, I used the following:

```php
use Filament\Pages\Actions;

// ...

    protected function getActions(): array
    {
        return [
            $this->getCancelFormAction(),
            $this->getSaveFormAction()
                ->action('save'),
        ];
    }

    protected function getFormActions(): array
    {
        return []; // necessary to remove the bottom actions
    }
```

**Comment by scottzirkel** [comment]
Is there a good solution to this problem in v3? Moving the actions into the getActions method moved them outside of the form and the getSaveFormAction no longer submits the form.

**Reply by lucacastelnuovo** [reply]
Unfortunately, I did not solve this yet. Placing the `save` button in the header actions breaks its functionality (It won't submit the form).

**Reply by RX-77IT** [reply]
What about `getHeaderActions()`?
[https://filamentphp.com/docs/3.x/panels/resources/editing-records#custom-actions](https://filamentphp.com/docs/3.x/panels/resources/editing-records#custom-actions)

NECRO-EDIT: now there is a proper entry in the v3 docs
[https://filamentphp.com/docs/3.x/panels/resources/editing-records#adding-a-save-action-button-to-the-header](https://filamentphp.com/docs/3.x/panels/resources/editing-records#adding-a-save-action-button-to-the-header)

**Reply by scottzirkel** [reply]
I had the same issues. Put it in the right place on the screen, but wrong place in the DOM.

**Reply by scottzirkel** [reply]
I will say, buttons like `Create & Continue Creating` work because they are not submitting the form, they are firing a Livewire event that runs the same method as submitting the form. While that works well, you still want to have a submit button that handles the submission formally. It might be doable if the save button had a `form="form-id"` on it, but I know that isn't fully functional in Livewire, so that might be hit or miss.

**Reply by freshleafmedia** [reply]
This is possible in v3 by doing the following in your Page:

```php
class EditExample extends \Filament\Resources\Pages\EditRecord
{
    protected static string $resource = ExampleResource::class;

    public function getFormActions(): array
    {
        return [];
    }

    protected function getHeaderActions(): array
    {
        return [
            $this->getSaveFormAction()
                ->submit(null)
                ->action(fn() => $this->save()),
            $this->getCancelFormAction(),
        ];
    }
}
```

**Reply by scottzirkel** [reply]
@freshleafmedia, not sure if you have some custom actions or not, but `getSaveFormAction()` is not a valid method. `getSubmitFormAction()` is, but it doesn't work for the reasons stated above.

I was able to get it to work, more or less. I had to create a custom view, fill it with the [default view contents](https://filamentphp.com/docs/3.x/panels/resources/creating-records#custom-view), and add the following to the `getHeaderActions` method.
 
```
CreateAction::make()
    ->label('Create')
    ->keyBindings(['enter', 'mod+s'])
    ->createAnother(false)
    ->action('create'),
```

@lucacastelnuovo I hope this helps!

**Reply by freshleafmedia** [reply]
@scottzirkel The method is defined here: https://github.com/filamentphp/filament/blob/870b856fe388c21541755fe06189537c55f08a60/packages/panels/src/Resources/Pages/EditRecord.php#L284

**Reply by scottzirkel** [reply]
@freshleafmedia Ah, I was looking in CreateRecord, where it does not exist. Thanks!
